include/maliput/math/kd_tree.h

```diff
@@ -0,0 +1,163 @@
+
+ private:
+  // Obtains squared distance between two points.
+  // @param point_a First point.
+  // @param point_b Second point.
+  // @returns The squared distance between the two points.
+  static double SquaredDistance(const Coordinate& point_a, const Coordinate& point_b);
+
+  // A node in the kd-tree.
+  // The node is in essence a point of the data structure that divides the upper parent node into two sub-trees, left
+  // and right.
+  class Node {
+   public:
+    // Constructs a Node.
+    // @param point The point that the node represents.
+    Node(const Coordinate& point) : point_(point) {}
+
+    // Returns the point that the node represents.
+    const Coordinate& get_point() const { return point_; }
+
+    // Obtains dimension's value.
+    // @param index Dimension to evaluate.
+    // @returns The value of the node in the @p index dimension.
+    //
+    // @throws maliput::common::assertion_error when @p index is greater than Dimension.
+    double get(std::size_t index) const;
+
+    // @returns The squared distance between the node and @p other point.
+    double Distance(const Coordinate& other) const { return SquaredDistance(point_, other); }
+
+    // Sets @p left as the left sub-node.
+    void set_left(Node* left) { left_ = left; }
+    // Sets @p right as the right sub-node.
+    void set_right(Node* right) { right_ = right; }
+    // @returns The left sub-node.
+    const Node* get_left() const { return left_; }
+    // @returns The right sub-node.
+    const Node* get_right() const { return right_; }
+
+   private:
+    Coordinate point_;
+    Node* left_{nullptr};
+    Node* right_{nullptr};
+  };
+
+  // Functor for comparing points according to the given dimension being evaluated at that point.
+  struct NodeCmp {
+    NodeCmp(std::size_t index);
+    // Compares two nodes according to the given dimension being evaluated at that point.
+    bool operator()(const Node& n1, const Node& n2) const;
+    std::size_t index_;
+  };
+
+  // Makes a tree from a range of points in the data structure.
+  // @param begin Is the start of range.
+  // @param end Is the end of range.
+  // @param index Is the dimension being evaluated.
+  Node* MakeTree(std::size_t begin, std::size_t end, std::size_t index);
+
+  // Obtains the nearest point in the @p node to the given @p point.
+  // The @p index under evaluation is provided as this method is called recursively.
+  // Updates the #visited, #best_, and #best_dist_ variables.
+  void Nearest(const Node* node, const Coordinate& point, std::size_t index) const;
+
+  // TODO(francocipollone): Add these NN-related variables to a struct.
+  //
+  // Best node found for nearest neighbour(NN) query.
+  mutable Node* best_ = nullptr;
+  // Best distance found for NN query.
+  mutable double best_dist_ = std::numeric_limits<double>::infinity();
+  // Number of nodes visited by the last call to Nearest().
+  mutable std::size_t visited_ = 0;
```

Comment:
I am not fond of these values being mutable. Can you instead update them as arguments in Nearest?

Reply:
Done.